src/protocol/sort/bit_permutation.rs

```diff
@@ -15,71 +15,57 @@ use futures::future::try_join_all;
 /// "An Efficient Secure Three-Party Sorting Protocol with an Honest Majority"
 /// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
 /// <https://eprint.iacr.org/2019/695.pdf>.
-#[derive(Debug)]
-pub struct BitPermutation<'a, F> {
-    input: &'a [Replicated<F>],
-}
 
-impl<'a, F: Field> BitPermutation<'a, F> {
-    pub fn new(input: &'a [Replicated<F>]) -> BitPermutation<'a, F> {
-        Self { input }
+#[embed_doc_image("bit_permutation", "images/sort/bit_permutations.png")]
+/// Protocol to compute a secret sharing of a permutation, after sorting on just one bit.
+///
+/// At a high level, the protocol works as follows:
+/// 1. Start with a list of `n` secret shares `[x_1]` ... `[x_n]` where each is a secret sharing of either zero or one.
+/// 2. Create a vector of length `2*n` where the first `n` rows have the values `[1 - x_1]` ... `[1 - x_n]`
+/// and the next `n` rows have the value `[x_1]` ... `[x_n]`
+/// 3. Compute a new vector of length `2*n` by computing the running sum of the vector from step 2.
+/// 4. Compute another vector of length `2*n` by multipling the vectors from steps 2 and 3 element-wise.
+/// 5. Compute the final output, a vector of length `n`. Each element `i` in this output vector is the sum of
+/// the elements at index `i` and `i+n` from the vector computed in step 4.
+///
+/// ![Bit Permutation steps][bit_permutation]
+/// ## Panics
+/// In case the function is unable to get double size of output from multiplication step, the code will panic
+///
+/// ## Errors
+/// It will propagate errors from multiplication protocol.
+pub async fn bit_permutation<F: Field>(
+    ctx: ProtocolContext<'_, Replicated<F>, F>,
+    input: Vec<Replicated<F>>,
```

Comment:
I thought passing the vector is better than a slice since the function can drop the input and hence take less memory? @martinthomson

Reply:
The general guidance for any API is to take values by reference (which is what a slice is) rather than by value, unless you need the value.  If the caller passes a slice and then subsequently drops that slice, that is effectively the same cost as passing by value.  However, if the caller needs to pass that same value twice, then asking it to provide a value means that it needs to make a copy of the data in order to make two calls.

Reply:
makes sense. Didnt know about this compiler optimization.